### PR TITLE
GTM: Add option to create a new account during setup

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #1080

## Relevant technical choices

* Used same language for creation instructions as was done for Analytics (which may be more recent than the AC?)  
Should we perhaps use a single translation string for both instances and use a placeholder for the service name? 
* Disabled Container select when account is not chosen (ie. after refreshing accounts)
* Replaced cryptic numeric magic values with explicit values for choose/create
* Added utilities for validating account and container IDs
* Simplified account refreshing without state
* Split request methods for accounts/containers or combined into dedicated methods
* Simplified conditionals when calling `setState` if `this._isMounted` with instance `setState` method  
Ideally we would fix the underlying events that cause errors when `setState` is called when the component is unmounted. This likely would require cancelling fetch requests that finish when the component is unmounting

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/tagmanager/setup.js [hunk 1]
```diff
@@ -68,6 +76,12 @@ class TagmanagerSetup extends Component {
 		this.refetchAccount = this.refetchAccount.bind( this );
 	}
 
+	setState() {
+		if ( this._isMounted ) {
+			Component.prototype.setState.apply( this, arguments );
+		}
+	}
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: This is quite hacky indeed. Let's instead cancel requests properly when the component is unmounted, as you already suggested.

File: assets/js/modules/tagmanager/setup.js [hunk 2]
```diff
@@ -189,51 +233,61 @@ class TagmanagerSetup extends Component {
...
-				errorCode = 'accountEmpty';
-				errorMsg = __(
-					'We didn’t find an associated Google Tag Manager account, would you like to set it up now? If you’ve just set up an account please re-fetch your account to sync it with Site Kit.',
-					'google-site-kit'
-				);
-			}
-
-			// Verify if user has access to the selected account.
-			if ( selectedAccount && ! accounts.find( ( account ) => account.accountId === selectedAccount ) ) { // Capitalization rule exception: `accountId` is a property of an API returned value.
-				data.invalidateCacheGroup( TYPE_MODULES, 'tagmanager', 'accounts-containers' );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: Just double-checking, why was this removed? I assume it's because we shouldn't cache this anyway?

File: assets/js/modules/tagmanager/setup.js [hunk 3]
```diff
@@ -189,51 +233,61 @@ class TagmanagerSetup extends Component {
 				usageContext,
 			};
 
-			let errorCode = false;
-			let errorMsg = '';
 			const { accounts, containers } = await data.get( TYPE_MODULES, 'tagmanager', 'accounts-containers', queryArgs );
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: Not for now, but I think we should abandon this datapoint in a future PR and instead only rely on `GET:accounts` and `GET:containers`.

File: assets/js/modules/tagmanager/setup.js [hunk 4]
```diff
@@ -297,72 +350,80 @@ class TagmanagerSetup extends Component {
 
 			global.googlesitekit.modules.tagmanager.settings = savedSettings;
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: This still makes me sad (I know it was here before) - can't wait for having a datastore.

File: assets/js/modules/tagmanager/setup.js [hunk 5]
```diff
@@ -297,72 +350,80 @@ class TagmanagerSetup extends Component {
...
+		this.setState( { selectedAccount: selectValue } );
+
+		if ( ACCOUNT_CHOOSE === selectValue ) {
 			this.setState( {
-				containersLoading: true,
-				selectedAccount: selectValue,
+				selectedContainer: ACCOUNT_CHOOSE,
+				containers: [ {
+					publicId: ACCOUNT_CHOOSE,
+					name: __( 'Select an account', 'google-site-kit' ),
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: This should say "Select a container". Having it say something different just because you haven't yet interacted with the other dropdown is confusing - the field being disabled is the only thing that should be different.

File: assets/js/modules/tagmanager/setup.js [hunk 6]
```diff
@@ -486,15 +544,20 @@ class TagmanagerSetup extends Component {
...
-							return (
-								<Option
-									key={ account.accountId /* Capitalization rule exception: `accountId` is a property of an API returned value. */ }
-									value={ account.accountId /* Capitalization rule exception: `accountId` is a property of an API returned value. */ }>
-									{ account.name }
-								</Option>
-							);
-						} ) }
+						{ []
+							.concat( ACCOUNT_CHOOSE === selectedAccount ? { accountId: ACCOUNT_CHOOSE, name: __( 'Select one...', 'google-site-kit' ) } : [] )
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: I think this should always be included, both for accounts and containers. Removing it once you've selected an account is unexpected, especially for accessibility.

File: assets/js/modules/tagmanager/setup.js [hunk 7]
```diff
@@ -503,20 +566,22 @@ class TagmanagerSetup extends Component {
...
-								name: __( 'Set up a new container', 'google-site-kit' ),
-								publicId: 0,
-							} ).map( ( { name, publicId }, i ) =>
-								<Option
-									key={ i }
-									value={ publicId /* Capitalization rule exception: `publicId` is a property of an API returned value. */ }>
-									{ name }
-								</Option>
-							) }
+							{ containers
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: Shouldn't the "Select a container" option be handled here like you do it for accounts, instead of actually adding it to the `containers` state var? I think that'd be cleaner.

File: assets/js/modules/tagmanager/setup.js (hunk 1 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: This may be less than ideal but it is much cleaner than before and I'm not sure if making requests cancellable is within the scope of this PR as it would likely require changes to our data API. Cancelling fetch requests requires the use of [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) which is technically an experimental API so I wasn't sure if we should rely on it just yet either.

File: assets/js/modules/tagmanager/setup.js (hunk 2 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: I believe this was added before we updated all data requests during setup to not be cached. Using cache is opt-in so this is redundant since all data requests do not set `nocache = false` as the fourth parameter.

File: assets/js/modules/tagmanager/setup.js (hunk 7 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Yes, agreed.

File: assets/js/modules/tagmanager/setup.js [hunk 8]
```diff
@@ -37,6 +37,15 @@ import { __, _x, sprintf } from '@wordpress/i18n';
...
 import { addFilter, removeFilter } from '@wordpress/hooks';
 
+/**
+ * Internal dependencies
+ */
+import { isValidAccountID, isValidContainerID } from './util';
+
+const ACCOUNT_CREATE = 'account_create';
+const ACCOUNT_CHOOSE = 'account_choose';
+const CONTAINER_CREATE = 'container_create';
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: I opted for a generic `UNSELECTED` named value of an empty string (since all choices are strings) 👍

File: assets/js/modules/tagmanager/setup.js (hunk 3 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: That would make it a bit slower to populate but that's probably a good idea for maintainability.

File: assets/js/modules/tagmanager/setup.js (hunk 5 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: This was borrowed from Analytics setup for consistency but that makes sense.

Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: When an option value is empty, the enhanced select only displays its label, even if there is an option with a matching value.

image

What do you think about using this instead of adding placeholder options? This seems to be more consistent with MDC behavior and results in less translation strings but maybe not as good for accessibility?

File: assets/js/modules/tagmanager/setup.js (hunk 3 above)
Comment (felixarntz) [reply assets/js/modules/tagmanager/setup.js]: Yeah good point, I thought about that too after writing this. Sounds cleaner, but a single route is indeed faster in certain circumstances, especially if we wanted to pre-select a container.

File: assets/js/modules/tagmanager/setup.js (hunk 5 above)
Comment (felixarntz) [reply assets/js/modules/tagmanager/setup.js]: I would go for the better accessibility. Having these placeholder options would also work in a regular `select`, so I think we should go for that.

File: assets/js/modules/tagmanager/setup.js [hunk 9]
```diff
@@ -486,37 +532,48 @@ class TagmanagerSetup extends Component {
...
-							} ).map( ( { name, publicId }, i ) =>
-								<Option
-									key={ i }
-									value={ publicId /* Capitalization rule exception: `publicId` is a property of an API returned value. */ }>
-									{ name }
-								</Option>
-							) }
+							{ []
+								.concat( containers )
+								.concat( {
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: We should probably add this option only if `! hasExistingTag`, like in the accounts dropdown.

File: assets/js/modules/tagmanager/setup.js [hunk 10]
```diff
@@ -486,37 +532,48 @@ class TagmanagerSetup extends Component {
...
-							return (
-								<Option
-									key={ account.accountId /* Capitalization rule exception: `accountId` is a property of an API returned value. */ }
-									value={ account.accountId /* Capitalization rule exception: `accountId` is a property of an API returned value. */ }>
-									{ account.name }
-								</Option>
-							);
-						} ) }
+						{ []
+							.concat( accounts )
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: As mentioned, we should include a "Select an account..." option as first one (placeholder option).

File: assets/js/modules/tagmanager/setup.js [hunk 11]
```diff
@@ -486,37 +532,48 @@ class TagmanagerSetup extends Component {
...
-								publicId: 0,
-							} ).map( ( { name, publicId }, i ) =>
-								<Option
-									key={ i }
-									value={ publicId /* Capitalization rule exception: `publicId` is a property of an API returned value. */ }>
-									{ name }
-								</Option>
-							) }
+							{ []
+								.concat( containers )
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: See above, we should include a "Select a container..." option as first one (placeholder option).

File: assets/js/modules/tagmanager/setup.js [hunk 12]
```diff
@@ -189,51 +225,61 @@ class TagmanagerSetup extends Component {
...
-					errorCode,
-					errorMsg,
-				} );
-			}
+			this.setState( {
+				isLoading: false,
+				accounts,
+				selectedAccount: selectedAccount || get( containers, [ 0, 'accountId' ] ), // Capitalization rule exception: `accountId` is a property of an API returned value.
+				containers,
+				selectedContainer: selectedContainer || get( containers, [ 0, 'publicId' ] ), // Capitalization rule exception: `publicId` is a property of an API returned value.
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: This shouldn't be addressed by this PR, but I think it's a bit strange that we just select the first account/container. I think we should revisit this on a broader scope as part of #1101 and following issues. Maybe only pre-select if
* either there is specific logic that determined this selection as likely (e.g. URL matching)
* or there is just one list entry (no point in selecting)

File: assets/js/modules/tagmanager/setup.js (hunk 9 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Done! 👍

File: assets/js/modules/tagmanager/setup.js (hunk 10 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Added back

File: assets/js/modules/tagmanager/setup.js (hunk 11 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Added back

File: assets/js/modules/tagmanager/setup.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: I had to give this a non-empty value going forward to work around the issue of it displaying the field label when the value is falsy.

[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/modules/tagmanager/setup.js]: Okay, I wasn't aware that the enhanced select displayed the field label inside if nothing is selected. In that case, I think the best we can do here to account for that is to get rid of our placeholders again and just rely on that. With that we stick closest to what Material recommends, and the only downside is that one cannot get back to the "there is nothing selected" state later. Which I guess is not too problematic.

For later, when we want to not preselect anything, the enhanced select's `value` prop should just be an empty string. This should be fine for accessibility.

Sorry for the back-and-forth here, let's just get rid of placeholders in enhanced select fields altogether then and rely on the default behavior of displaying the label.

File: assets/js/modules/tagmanager/setup.js (hunk 10 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Removed again 😄

File: assets/js/modules/tagmanager/setup.js (hunk 11 above)
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: Removed again 😄

File: assets/js/modules/tagmanager/setup.js [hunk 13]
```diff
@@ -51,7 +60,6 @@ class TagmanagerSetup extends Component {
 			containers: [],
 			errorCode: false,
 			errorMsg: '',
-			refetch: false,
 			selectedAccount: settings.accountID,
 			selectedContainer: settings[ containerKey ],
```
Comment (felixarntz) [inline assets/js/modules/tagmanager/setup.js]: See here. This could potentially make the value `undefined`. It never should for us because these values coming from PHP are strings, but still a bit error-prone.

I think we should get rid of the specific empty string value and the `UNSELECTED` constant again and just check `! value` instead.

File: assets/js/modules/tagmanager/setup.js [hunk 14]
```diff
@@ -486,37 +532,48 @@ class TagmanagerSetup extends Component {
...
 
 					{ containersLoading ? ( <ProgressBar small /> ) : (
 						<Select
+							className="googlesitekit-tagmanager__select-container"
 							enhanced
 							name="containers"
 							label={ __( 'Container', 'google-site-kit' ) }
 							value={ selectedContainer }
-							disabled={ hasExistingTag }
+							disabled={ hasExistingTag || [ ACCOUNT_CREATE, UNSELECTED ].includes( selectedAccount ) }
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/modules/tagmanager/setup.js]: I simplified the changed portion to ` || ! isValidAccountID( selectedAccount )`
